docs/CONFIG.md
# Privacy Pass configuration

Privacy Pass uses cryptographic tokens to bypass internet challenges
from certain providers. In this doc, we discuss the configuration file
that Privacy Pass uses for interpreting when tokens should be sent for
signing/redemption. See `src/ext/config.js` for an example configuration
`exampleConfig` describing the correct format of the JSON struct.

## config.js

Holds the various configurations as JavaScript JSON structs for the
providers that Privacy Pass interacts with. Currently, only the
Cloudflare and hCaptcha configs are active. There is an example config.
In the following we will highlight how each config field is used.

### config\["id"\]

A unique identifier highlighting which config is used. Currently
`cfConfig.id = 1` and `hcConfig.id = 2`, all other configs must be added
with unique "id" values.

### config\["sign"\]

A bool dictating whether Privacy Pass should send tokens for signing.

### config\["redeem"\]

A bool dictating whether Privacy Pass should send tokens for redemption.

### config\["max-spends"\]

The integer number of tokens that should be redeemed per host in each
interaction. This prevents Privacy Pass from repeatedly spending tokens
for the same host (if some unknown issues occur, for example).

### config\["max-tokens"\]

The integer number of tokens that should be held by Privacy Pass at any
one time.

### config\["var-reset"\]

A bool dictating whether the set of variables holding previous
redemption information should be reset (see resetVars() in
background.js). We recommend this is set to true, since these variables
prevent tokens from being spent in the future.

### config\["var-reset-ms"\]

The time intervals (ms) by which the variables from above are reset.

### config\["comm-vk"\]

Public verification key, used to verify signatures on commitments that
are retrieved over HTTP.

### config\["commitments"\]

Hex-encoded elliptic curve commitment values for verifying DLEQ proofs.
These essentially amount to public keys issued by the provider. The
values of "G" and "H" held in `config["commitments"]["dev"]` should be
used for development purposes. Those held in
`config["commitments"][<version>]` should be used in the production
environment.

### config\["spending-restrictions"\]

A JSON struct of restrictions for redeeming tokens

#### config\["spending-restrictions"\]\["status-code"\]

An integer corresponding to the status code returned by the server. This
HTTP status code is checked before a redemption is initiated, for
Cloudflare this value is set to 403. That is, token redemptions can only
occur after a HTTP response with status code 403 is received. For
hCaptcha the code is 200, however only hCaptcha specific URLs are
eligible for the spend.

#### config\["spending-restrictions"\]\["max-redirects"\]

An integer dictating the number of times that tokens should be spent
after requests have been redirected. That is, consider a HTTP response
that Privacy Pass deems suitable to initiate a redemption. This number
indicates how may redemption HTTP requests will be tolerated where the
response from the server results in HTTP redirection.

#### config\["spending-restrictions"\]\["new-tabs"\]

An array of strings indicating that the tab that is open corresponds to
a new tab (and thus token redemption should not occur).

#### config\["spending-restrictions"\]\["bad-navigation"\]

An array of strings corresponding to chrome.webNavigation methods that
indicate navigation types where tokens should not be redeemed. In the
case of Cloudflare, this is limited to `auto_subframe` navigations that
are not used for Cloudflare CAPTCHAs.

#### config\["spending-restrictions"\]\["bad-transition"\]

Similar to above, except for transition types. For Cloudflare, we rule
out redemption requests when `server_redirect` is the transition type.

#### config\["spending-restrictions"\]\["valid-redirects"\]

An array of strings indicating the URL redirections that are tolerated
when tokens are being redeemed. For example, redemptions that upgrade
HTTP connections to HTTPS connections.

#### config\["spending-restrictions"\]\["valid-transitions"\]

An array of strings indicating the transition types that are definitely
valid, when considering whether redemption requests should be
sanctioned.

### config\["spend-action"\]

A JSON struct of configuration related to redeeming tokens.

#### config\["spend-action"\]\["urls"\]

URLs that activate WebRequest listeners for redemption, "`<all_urls>`"
corresponds to matching all possible URLs. All the GET query strings will
be ignored.

#### config\["spend-action"\]\["redeem-method"\]

A string that determines the method that token redemptions are handled.
Currently the only supported methods are `"reload"` and `"no-reload"`.
`"reload"` method redeems tokens by reloading the page and appending
tokens to the subsequent HTTP request. `"no-reload"` method sends tokens
with the initial HTTP request that is supposed to fetch the captcha
itself.

#### config\["spend-action"\]\["header-name"\]

The name of the header that contains a token for redemption.

#### config\["spend-action"\]\["header-host-name"\]

The name of the header that contains a hostname that is sending the
redemption request.

#### config\["spend-action"\]\["header-path-name"\]

The name of the header that contains a path that is sending the
redemption request.

#### config\["spend-action"\]\["empty-resp-headers"\]

If an empty set of response headers is received with the correct status
code (as defined in `config["spending-restrictions"]["status-code"]`)
then this object contains an array of strings that correspond to
possible ways of acquiring the headers. Currently we only support
`"direct-request"`, which sends a direct request to the same URL with a
specific endpoint attached as defined in
`config["opt-endpoints"]["challenge"]`.

This option was introduced to mitigate problems with Chrome in
conjunction with sub-resources hosted on separate Cloudflare domains.

### config\["issue-action"\]

A JSON struct of configuration related to issuing tokens.

#### config\["issue-action"\]\["urls"\]

URLs that activate WebRequest listeners for issuance, "`<all_urls>`"
corresponds to matching all possible URLs. All the GET query strings will
be ignored.

#### config\["issue-action"\]\["sign-reload"\]

A bool dictating whether the page should be reloaded after tokens are
successfully signed.

#### config\["issue-action"\]\["sign-resp-format"\]

Format of the response (as a string) to a signing request. Currently,
support "string" or "json". When "string" is used, expect the signed
tokens to be included (base-64 encoded) in the HTTP response body in the
form `signatures= || <signed-tokens> || <Batch-DLEQ-Resp>`. When "json"
is used, expect the signed tokens to be included (base-64 encoded) as a
JSON struct with key: "signatures" and value `<signed-tokens> ||
<Batch-DLEQ-Resp>`.

#### config\["issue-action"\]\["tokens-per-request"\]

The integer number of tokens that should be sent with each signing
request. For Cloudflare, there is also a server-side upper bound of 100
tokens for each signing request. We recommend that this is enforced to
prevent unlimited numbers of tokens being signed.

#### config\["issue-action"\]\["request-identifiers"\]

An object containing optional identifying characteristics of requests.
Requests identified using these characteristics can be augmented with
data required for the issuance portion of the Privacy Pass protocol.

For example, the CF config (id=1) uses the following parameters:

- `"query-param"`: for identifying requests based on the URL query
  parameters
- `"body-param"`: for replaying important parameters that part of the
  request body
- `"post-processed"`: for identifying requests that have already been
  processed by the WebRequest API

### config\["cookies"\]

A JSON struct of configuration related to cookie management.

#### config\["cookies"\]\["check-cookies"\]

A boolean value that determines cookies should be checked before tokens
are sent for redemption. That is, a token is not redeemed if the browser
has a clearance cookie for the URL that redemption occurring for.

#### config\["cookies"\]\["clearance-cookie"\]

A string that specifies the specific name of the type of clearance
cookie used by the provider. In the case of Cloudflare, this is
`"cf_clearance"`.

### config\["captcha-domain"\]

A string specifying a domain where users can obtain signed tokens by
solving a challenge/CAPTCHA. This is helpful to allow users to build up
initial stockpiles of tokens before they browse.

### config\["get-more-passes-url"\]

A valid URL string specifying an URL where users can get more signed
tokens by solving a challenge/CAPTCHA. This URL is pulled into the
pop-up `Get more passes` menu item. Can be different from
`captcha-domain`, and must be a valid URL with protocol specified.

### config\["opt-endpoints"\]

Optional endpoints for use by the particular configuration.

### config\["error-codes"\]

A JSON struct of configuration related to error codes.

#### config\["error-codes"\]\["verify-error"\]

String error code that the server returns if token redemption fails due
to a signature verification error.

#### config\["error-codes"\]\["connection-error"\]

String error code that the server returns if an internal connection
error occurs server-side.

### config\["h2c-params"\]

A JSON struct of parameters for the curve setting that the client uses.
These settings are sent to the server with all redemption requests.

#### config\["h2c-params"\]\["curve"\]

A string that defines the elliptic curve that is being used for mapping
tokens to. Currently we only support the NIST curve P256, this is
represented by the string "p256".

#### config\["h2c-params"\]\["hash"\]

The actual hash function that is used for hashing bytes to the base
field of the curve. Currently we only support the usage of the SHA256
hash function, using the string "sha256".

#### config\["h2c-params"\]\["method"\]

The hash-to-curve method that is used for moving field elements to curve
points. If it is set to "increment" then we use the hash-and-increment
method that will be deprecated moving forward. If it is set to "swu",
then we use the affine version of the SWU algorithm implemented in
h2c.js (see [HASH_TO_CURVE](HASH_TO_CURVE.md) for a
description of the algorithm in full).

### config\["send-h2c-params"\]

A boolean that determines whether the contents of `config["h2c-params"]`
should actually be sent to the server.

## Patching configurations

Configurations can be patched by adding patches to the configuration
available at <https://github.com/privacypass/ec-commitments>. This
allows external configuration changes to be made without requiring an
extension update. A patch blob must be signed with the secret key
corresponding to the public available at `config["comm-vk"]`.

Previously, the configuration available at the URL above looks something
like:

```json
"CF": {
    "1.0": {
        // ...
    },
    "1.01": {
        // ...
    },
    // more ...
}
```

Adding patches to the `"CF"` would require adding a `"patches"` field to
take the following form:

```json
"CF": {
    "patches": [
        {
            "min-version": "2.0.0",
            "config": {
                // example patched fields
                "captcha-domain": "new-domain",
                "spend-action": {
                    "urls": ["https://some.example.com"],
                    "header-name": "new-header",
                },
                // more fields ...
            },
            "sig": "<signature_value>"
        },
        {
            "min-version": "x.y.z",
            "config": {
                // patched fields
            },
            "sig": "<signature_value>"
        }
    ],
    "1.0": {
        // ...
    },
    "1.01": {
        // ...
    },
    ...
}
```

In other words, `patches` is an array containing multiple configuration
patches, each corresponding to a minimum version of the extension and a
signature. The `sig` value is an ECDSA signature computed over the
contents of the `config` and `min-version` members. When patches are
specified in the external configuration, they must come with a valid
signature to be accepted as modifications to the base config specified
in the extension.

Patches are only permitted for the following configuration members:
```
[
    "max-spends",
    "max-tokens",
    "var-reset",
    "var-reset-ms",
    "spending-restrictions",
    "spend-action",
    "issue-action",
    "cookies",
    "captcha-domain",
    "get-more-passes-url",
    "opt-endpoints",
    "error-codes",
    "h2c-params",
    "send-h2c-params",
]
```
Patches for other members will be rejected.

For further examples of how to patch configurations, see
`config.test.js`.


src/ext/config.js
/*
* Config file for handling DLEQ proofs from edge
*
* @author: Alex Davidson
*/
/* exported DevCommitmentConfig */
/* exported CHL_BYPASS_SUPPORT */
/* exported CHL_BYPASS_RESPONSE */
/* exported CONFIGURATION_URL */
/* exported PPConfigs */
/* exported processConfigPatches */
/* exported validRedemptionMethods */
/* exported retrieveConfiguration */
/* exported extVersion */

/**
 * Returns the version of the extension that is currently running
 * @return {string} Extension version
 */
function extVersion() {
    return chrome.runtime.getManifest().version;
}

/**
 * Returns the version of the extension as a separated array
 * @return {Array<Number>} Extension version
 */
function extVersionAsArray() {
    return versionStringAsNumbers(extVersion());
}

const CHL_BYPASS_SUPPORT = "cf-chl-bypass"; // header from server to indicate that Privacy Pass is supported
const CHL_BYPASS_RESPONSE = "cf-chl-bypass-resp"; // response header from server, e.g. with erorr code
const validRedemptionMethods = () => ["reload", "no-reload"]; // specifies valid token redemption methods
const CONFIGURATION_URL = "https://raw.githubusercontent.com/privacypass/ec-commitments/master/commitments-p256.json";

// initialise configurations from base settings, and potentially modify
// with patches later
let VALID_CONFIGS = [
    exampleConfig(),
    cfBaseConfig(),
    hcBaseConfig(),
];

let PPConfigs = () => Object.assign([], VALID_CONFIGS);

/**
 * Generates exampleConfig configuration object
 * @return {Object} Example configuration object
 */
function exampleConfig() {
    return {
        "id": 0, // unique integer identifying each individual config
        "name": "example", // string identifier of the configuration
        "long-name": "example", // full name of config
        "dev": true, // sets whether the configuration should only be used in development
        "sign": true, // sets whether tokens should be sent for signing
        "redeem": true, // sets whether tokens should be sent for redemption
        "max-spends": 1, // for each host header, sets the max number of tokens that will be spent, undefined for unlimited
        "max-tokens": 10, // max number of tokens held by the extension
        "var-reset": true, // whether variables should be reset after time limit expires
        "var-reset-ms": 100, // variable reset time limit
        "comm-vk":
            "-----BEGIN PUBLIC KEY-----\n" +
            "(PEM)\n" +
            "-----END PUBLIC KEY-----", // a PEM-encoded public key for ecdsa P-256.
        "commitments": { // an optional set of commitments that are specified inside the extension
            "1.0": {
                "G": "",
                "H": "",
            },
        },
        "spending-restrictions": {
            "status-code": [200], // array of status codes that should trigger token redemption (e.g. 403 for CF)
            "max-redirects": 2, // when page redirects occur, sets the max number of redirects that tokens will be spent on
            "new-tabs": ["about:privatebrowsing", "chrome://", "about:blank"], // urls that should not trigger page reloads/redemptions (these should probably be standard)
            "bad-navigation": ["auto_subframe"], // navigation types that should not trigger page reloads/redemptions (see: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/webNavigation/TransitionType)
            "bad-transition": ["server_redirect"], // transition types that should not trigger page reloads/redemptions (see: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/webNavigation/TransitionType)
            "valid-redirects": ["https://", "https://www.", "http://www."], // valid redirects that should trigger token redemptions
            "valid-transitions": ["link", "typed", "auto_bookmark", "reload"], // transition types that fine for triggering redemptions (see: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/webNavigation/TransitionType)
        }, // These spending restrictions are examples that apply in the CF case
        "spend-action": {
            "urls": ["<all_urls>"], // urls that listeners act on
            "redeem-method": "", // what method to use to perform redemption, currently we support "reload" for CF and "no-reload" for HC.
            "header-name": "challenge-bypass-token", // name of header for sending redemption token
            "header-host-name": "challenge-bypass-host", // needed for no-reload method
            "header-path-name": "challenge-bypass-path", // needed for no-reload method
            "empty-resp-headers": [], // if a HTTP response returns with no headers, specify what action to take; default is no action, also support "direct-request"
        },
        "issue-action": {
            "urls": ["<all_urls>"],
            "sign-reload": true, // whether pages should be reloaded after signing tokens (e.g. to immediately redeem a token)
            "sign-resp-format": "string", // formatting of response to sign request (string or json)
            "tokens-per-request": 5, // number of tokens sent for each signing request (e.g. 30 for CF)
            "request-identifiers": { // parameters used to identify requests that issuance data should be included into
                "query-param": "", // request identifier in query params of URL
                "body-param": [""], // request identifier array for checking of HTTP request body and adding to request
                "post-processed": "", // identifier for requests that have already been processed by WebRequest API
            },
        },
        "cookies": {
            "check-cookies": true, // whether cookies should be checked before spending
            "clearance-cookie": "", // name of clearance cookies for checking (cookies that are optionally acquired after redemption occurs)
        },
        "captcha-domain": "", // optional domain for acquiring tokens
        "get-more-passes-url": "", // optional url that the Get More Passes menu item will point, must be valid URL with protocol
        "opt-endpoints": {}, // optional endpoints for integration-specific operations
        "error-codes": {
            "connection-error": "5", // error code sent by server for connection error
            "verify-error": "6", // error code sent by server for verification error
            "bad-request-error": "7", // error code sent by server for signalling that a bad request was made
            "unknown-error": "8", // error code sent by server in case of an unknown error occurring
        }, // generic error codes (can add more)
        "h2c-params": {// parameters for establishing which hash-to-curve setting the client wants to use
            "curve": "p256", // elliptic curve that generated tokens should be mapped to; "p256", "p384", "p521"
            "hash": "sha256", // hash function for mapping bytes to base-field of elliptic curve; "sha256", "sha512"
            "method": "increment", // specifies which hash-to-curve method we should use; "increment" = hash-and-increment (the original but deprecated method); "swu" = optimised affine SWU algorithm (new method)
        },
        "send-h2c-params": false, // specifies whether to send the additional h2c-params with issue requests
    };
}

/**
* Returns the base configuration (without patches) used to integrated
* with Cloudflare
* @return {Object} base JSON configuration
 */
function cfBaseConfig() {
    const cfDomain = "captcha.website";
    const cfConfig = exampleConfig();
    cfConfig.id = 1;
    cfConfig.dev = false;
    cfConfig.name = "CF";
    cfConfig["long-name"] = "Cloudflare";
    cfConfig["max-tokens"] = 300;
    cfConfig["max-spends"] = 2;
    cfConfig["var-reset-ms"] = 2000;
    cfConfig["comm-vk"] =
        "-----BEGIN PUBLIC KEY-----\n" +
        "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAExf0AftemLr0YSz5odoj3eJv6SkOF\n" +
        "VcH7NNb2xwdEz6Pxm44tvovEl/E+si8hdIDVg1Ys+cbaWwP0jYJW3ygv+Q==\n" +
        "-----END PUBLIC KEY-----";
    cfConfig["spending-restrictions"]["status-code"] = [403];
    cfConfig["spend-action"]["redeem-method"] = "reload";
    cfConfig["issue-action"]["tokens-per-request"] = 30;
    cfConfig["issue-action"]["request-identifiers"]["query-param"] = "__cf_chl_captcha_tk__";
    cfConfig["issue-action"]["request-identifiers"]["body-param"] = ["g-recaptcha-response", "h-captcha-response", "cf_captcha_kind"];
    cfConfig["issue-action"]["request-identifiers"]["post-processed"] = "captcha-bypass";
    cfConfig.cookies["clearance-cookie"] = "cf_clearance";
    cfConfig["captcha-domain"] = cfDomain;
    cfConfig["get-more-passes-url"] = `https://${cfDomain}`;
    cfConfig["send-h2c-params"] = true;
    cfConfig["opt-endpoints"].challenge = "/cdn-cgi/challenge";
    cfConfig["spend-action"]["empty-resp-headers"] = ["direct-request"];
    // old version 1.0 commitments for backwards compatibility
    cfConfig["commitments"]["1.0"]["G"] = "BOidEuO9HSJsMZYE/Pfc5D+0ELn0bqhjEef2O0u+KAw3fPMHHXtVlEBvYjE5I/ONf9SyTFSkH3mLNHkS06Du6hQ=";
    cfConfig["commitments"]["1.0"]["H"] = "BHOPNAWXRi4r/NEptOiLOp8MSwcX0vHrVDRXv16Jnowc1eXXo5xFFKIOI6mUp8k9/eca5VY07dBhAe8QfR/FSRY=";
    return cfConfig;
}

/**
* Returns the base configuration (without patches) used to integrated
* with hCaptcha
* @return {Object} base JSON configuration
 */
function hcBaseConfig() {
    const hcConfig = exampleConfig();
    hcConfig.id = 2;
    hcConfig.dev = false;
    hcConfig.name = "HC";
    hcConfig["long-name"] = "hCaptcha";
    hcConfig["max-spends"] = undefined;
    hcConfig["max-tokens"] = 300;
    hcConfig["var-reset-ms"] = 2000;
    hcConfig["comm-vk"] =
        "-----BEGIN PUBLIC KEY-----\n" +
        "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE4OifvSTxGcy3T/yac6LVugArFb89\n" +
        "wvqGivp0/54wgeyWkvUZiUdlbIQF7BuGeO9C4sx4nHkpAgRfvd8jdBGz9g==\n" +
        "-----END PUBLIC KEY-----";
    hcConfig["spending-restrictions"]["status-code"] = [200];
    hcConfig["spend-action"]["redeem-method"] = "no-reload";
    hcConfig["spend-action"]["urls"] = ["https://*.hcaptcha.com/getcaptcha", "https://*.hmt.ai/getcaptcha", "http://127.0.0.1/getcaptcha"];
    hcConfig["issue-action"]["urls"] = ["https://*.hcaptcha.com/checkcaptcha/*", "https://*.hmt.ai/checkcaptcha/*", "http://127.0.0.1/checkcaptcha/*"];
    hcConfig["issue-action"]["sign-reload"] = false;
    hcConfig["issue-action"]["sign-resp-format"] = "json";
    hcConfig.cookies["clearance-cookie"] = "hc_clearance";
    hcConfig["captcha-domain"] = null;
    hcConfig["get-more-passes-url"] = "https://www.hcaptcha.com/privacy-pass";
    hcConfig["send-h2c-params"] = true;
    hcConfig["commitments"] = {
        "1.0": {
            "G": "BMKCnVDWUEBNiyAR+p0YT7QvtrOfpHAeatzipwo6x98Ch1q3ZoCkNdiQvUTEwDzG20RplG/IE2NCpsXZGLsUdvA=",
            "H": "BNJIpofS4RhbUfnkblr5yvuymaEfV+ViKshsoN9DkCRaHBB+TiKUnicc14gBswpLfBaKXuC102Cvwzq3YIN8dVo=",
        },
    };
    return hcConfig;
}

// Top-level configuration keys that can be patched externally
const PATCHABLE_KEYS = [
    "max-spends",
    "max-tokens",
    "var-reset",
    "var-reset-ms",
    "spending-restrictions",
    "spend-action",
    "issue-action",
    "cookies",
    "captcha-domain",
    "get-more-passes-url",
    "opt-endpoints",
    "error-codes",
    "h2c-params",
    "send-h2c-params",
];

/**
 * Processes valid patches to the available configurations. This
 * function is run in init.js
 * @param {Number} cfgId ID of the configuration being used
 * @return {XMLHttpRequest} the XHR used to retrieve any patches
 */
function processConfigPatches(cfgId) {
    const callback = (retrieved) => {
        const patches = retrieved["patches"];
        if (!patches) {
            return;
        } else if (!(patches instanceof Array)) {
            console.warn("[privacy-pass]: Patches not specified in correct format");
            return;
        }

        // process patches if they verify correctly
        patches.forEach((patch) => {
            if (!patch || typeof patch !== "object") {
                console.warn("[privacy-pass]: Patch not specified in correct format");
                return;
            }

            // try to verify patch
            if (!applicablePatch(patch["min-version"])) {
                console.warn("[privacy-pass]: Patch version criteria not met. Not applying following: " + JSON.stringify(patch, null, 4));
                return;
            }

            // check if signature is valid
            if (patch["sig"] === undefined) {
                console.warn("[privacy-pass]: Signature field for patch is missing not processing them");
                return;
            }

            // error is thrown if bad verification, just ignore if it
            // fails
            try {
                verifyConfiguration(cfgId, patch);
            } catch (e) {
                console.warn("[privacy-pass]: Not processing patch as unable to verify signature");
                return;
            }

            const config = getConfigForId(cfgId);
            const patchConfig = patch["config"];
            Object.keys(patchConfig).forEach((key) => {
                if (!PATCHABLE_KEYS.includes(key)) {
                    // do not process patch for non-patchable fields
                    console.warn(`[privacy-pass]: Patches for ${key} are not permitted`);
                    return;
                }
                const patchValue = patchConfig[key];
                let current = config[key];
                switch (typeof current) {
                    case "object":
                        if (current !== null) {
                            Object.assign(current, patchValue);
                        } else {
                            current = patchValue;
                        }
                        break;
                    default:
                        current = patchValue;
                        break;
                }
                config[key] = current;
            });

            // modify valid configs
            VALID_CONFIGS[cfgId] = config;
        });
    };
    const xhr = retrieveConfiguration(cfgId, callback);
    xhr.send();
    return xhr;
}

/**
 * Creates an XMLHttpRequest object to retrieve the active JSON
 * configuration for the given provider
 * @param {Number} cfgId ID of configuration to retrieve
 * @param {Function} callback function to execute after configuration is
 * retrieved
 * @return {XMLHttpRequest}
 */
function retrieveConfiguration(cfgId, callback) {
    if (!callback) {
        throw new Error("[privacy-pass]: Invalid configuration retrieval callback specified");
    }
    const xhr = new XMLHttpRequest();
    xhr.open("GET", CONFIGURATION_URL, true);
    xhr.setRequestHeader("Accept", "application/json");
    xhr.onreadystatechange = function() {
        if (xhrGoodStatus(xhr.status) && xhrDone(xhr.readyState)) {
            const provider = getConfigName(cfgId);
            const resp = JSON.parse(xhr.responseText);
            const config = resp[provider];
            callback(config);
        }
    };
    return xhr;
}

/**
 * Returns the version number (string of form x.y.z) as an array of
 * three numbers to make comparisons easier
 * @param {string} version
 * @return {Array<Number>}
 */
function versionStringAsNumbers(version) {
    return version.split(".").map((s) => parseInt(s));
}

/**
 * Compares the version of a patch with that of the extension and
 * returns true if the min patch version is satisfied.
 * @param {string} pv Patch version to compare with extension version
 * @return {boolean}
 */
function applicablePatch(pv) {
    if (!pv) {
        console.warn("[privacy-pass]: No version specified in patch.");
        return false;
    }

    let arr;
    try {
        arr = versionStringAsNumbers(pv);
    } catch (e) {
        console.warn("[privacy-pass]: Failed to parse patch version.");
        return false;
    }

    // compare version strings
    const extVersion = extVersionAsArray();
    if (extVersion[0] < arr[0]) {
        return false;
    } else if (extVersion[0] === arr[0]) {
        if (extVersion[1] < arr[1]) {
            return false;
        } else if (extVersion[1] === arr[1]) {
            if (extVersion[2] < arr[2]) {
                return false;
            } else if (
                // typically versions consist of 3 numbers, but Chrome
                // supports up to 4 so we should check for this. If the
                // lengths differ and they are equal up to this point
                // then just return true.
                extVersion.length === 4
                && arr.length === 4
                && extVersion[2] === arr[2]
            ) {
                if (extVersion[3] < arr[3]) {
                    return false;
                }
            }
        }
    }
    return true;
}
